docs/rfcs/20220131_raft_snapshots.md

```diff
@@ -0,0 +1,149 @@
+
+![](./imgs/20220131_cleanup.png)
+
+Imagine that Redpanda stones withdraw & deposit history for a set of accounts and an aggregate-dependent operation needs an account balance.
+
+![](./imgs/20220131_compaction.png)
+
+We see that the state after compaction is completely different: `(2,-2,4)` vs `(3,1,4)`. The only way to guarantee the correct result is to cache state before the compaction / deletion.
+
+Local caching isn't enough. Raft tolerates note failures as long as majority is connected meaning a single node may fall behind the majority and behind the compaction horizon.
+
+![](./imgs/20220131_behind.png)
+
+When the isolated node rejoins the cluster it fetches an already compacted log with holes and the state diverges.
+
+# Reference-level explanation
+
+## Approach #1: Ship the snapshot (cached state) via the log
+
+The main idea is to use existing log replication to ship the snapshot. It's easier to implement but just like the reconfiguration batches it shifts the offsets and makes us continue to rely on offset translation.
+
+### Overview
+
+A leader decides to do compaction at offset X:
+ 
+  1. saves current snapshot
+  2. waits until it's safe to do compaction:
+    a. segment containing X is closed
+    b. tx front is passed the end of the segment
+  3. replicates snapshot
+ 
+A leader or a follower sees the snapshot via `state_machine::apply` (it happens only when the batch is replicated to the majority and can't be lost):
+
+  - perform the compaction when current node is the leader and it sent the snapshot (within current term)
+  - when the snapshot isn't applied:
+    1. rewind log at offset X
+    2. apply the snapshot
+  - when the snapshot is applied:
+    - perform the compaction
+
+There is a problem with this approach: a follower doesn't know if the batches it replays contain holes so its current state may be incorrect until it sees the snapshot batch. This problem becomes actual when Redpanda supports reads from replicas. There are several ways to deal with it.
+
+#### Blocking replica reads
+
+A follower is allowed to serve reads only up to the last observed snapshot. Even when the prior batches contain holes they become visible when the snapshot is applied and corrected.
+
+#### Modify the `append_entries_request`
+
+We may modify the append_enties_request to include an offset of the last compaction. When a follower sees that last compaction less than current offset it knows that the latest batches don't have holes and it's safe to serve the requests.
+
+When we modify `append_entries_request` or introduce and switch to the new API we need to thing of the update story. In order to avoid shutting down a cluster for the update we need to put the change behind a feature flag, update the brokers one by one and only then to active the feature.
```

Comment:
```suggestion
When we modify `append_entries_request` or when we introduce and switch to the new API, we need to take upgrades into account. In order to avoid shutting down a cluster for an update, we need to put the change behind a feature flag, update the brokers one by one, and only then active the feature.
```